File: hopfield/static/hopfield.js
NEURONS = 36
/*
http://stackoverflow.com/questions/6274339/how-can-i-shuffle-an-array-in-javascript
Jonas Raoni Soares Silva
http://jsfromhell.com/array/shuffle [v1.0]
*/
shuffle = function(o){ //v1.0
    for(var j, x, i = o.length; i; j = parseInt(Math.random() * i), x = o[--i], o[i] = o[j], o[j] = x);
    return o;
};

  var compare = function(a, b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < b.length; i++) {
        if (a[i].compare) { 
            if (!a[i].compare(b[i])) return false;
        }
        if (a[i] !== b[i]) return false;
    }
    return true;
}

var calculate_weight = function(i,j, patterns) {
    total = 0
    for (var p in patterns) {
        var pattern = patterns[p];
        total += (2*pattern[i] -1)*(2*pattern[j] -1);
    }
    return total;
};

var calculate_weights = function(patterns, neurons) {
    var weights = new Array(neurons);
    for (var i = 0; i < neurons; i++) {
        weights[i] = new Array(neurons);
        for (var j = 0; j < neurons; j++) {
            if (i == j) { 
              weights[i][j] = 0;
            } else {
              weights[i][j] = calculate_weight(i, j, patterns);
            }
        }
    }
    return weights;
};


var calculate_weighted_inputs = function(state, node, weights, num_neurons) {
    total = 0
    for (var i = 0; i < num_neurons;i++) { //for each neuron
        if (i != node) {
            total += weights[i][node] * state[i];
        }
    }
    return total;
}
var update_net = function(initial_state, weights, num_neurons) {
    has_changed = true;
    var order = new Array(num_neurons);
    for (var i = 0; i < num_neurons;i++)order[i] =i;
    order = shuffle(order)
    //order =[21, 26, 32, 17, 33, 19, 31, 13, 11, 7, 14, 16, 5, 20, 25, 29, 0, 27, 2, 6, 23, 9, 28, 34, 30, 15, 1, 4, 24, 18, 8, 12, 10, 22, 35, 3];
    last_state = initial_state.slice();
    while (has_changed) {
        has_changed = false;
        for (i in order) {
            current_node = order[i];
            value = calculate_weighted_inputs(initial_state, current_node, weights, num_neurons);
            if (value >= 0) initial_state[current_node] = 1;
            else initial_state[current_node] = 0;
        }
        if (!compare(initial_state, last_state)) {
            has_changed = true;
            last_state = initial_state.slice();
        }
    }
    return initial_state;
}

File: hopfield/hopfield.py
NEURONS = 36
import random
order = range(0,NEURONS)
random.shuffle(order)
#order = [2, 1, 0, 4, 3]
def calculate_weights(patterns):
    weights = [[0 for i in range(NEURONS)] for j in range(NEURONS)]
    for i in range(NEURONS):
        for j in range(NEURONS):
            if i == j: continue
            weights[i][j] = calculate_weight(i, j, patterns)
    print "memorizados:\n"
    for pattern in patterns:
        print_matrix(pattern)
        print "\n"
    return weights

def calculate_weight(i,j, patterns):
    total = 0
    for pattern in patterns:
        total += (2*pattern[i] -1)*(2*pattern[j] -1)
    return total

def calculate_weighted_inputs(state, node, weights):
    total = 0
    for i in range(NEURONS):
        if i != node:
            total += weights[i][node] * state[i]
    return total

def update_net(initial_state, weights):
    has_changed = True
    last_state = initial_state[:]
    print "entra"
    print print_matrix(initial_state)
    while has_changed:
        has_changed = False
        for current_node in order:
            value = calculate_weighted_inputs(initial_state, current_node, weights)
            if value >= 0: initial_state[current_node] = 1
            else: initial_state[current_node] = 0
        if initial_state != last_state:
            has_changed = True
            last_state = initial_state[:]
            #  initial_state[current_node] = 1
            #else: initial_state[current_node] = 0
    print "encontrado:"
    print print_matrix(initial_state)
    return initial_state

def print_matrix(matrix):
  li = ""
  lenght = len(matrix) / 10
  for i in range(6):
      li = ""
      for j in range(6):
          li = li + str(matrix[(i * 6) + j])
      print li


def print_matrix2(matrix):
  li = ""
  for i in range(6):
      li = ""
      for j in range(6):
          li = li + str(matrix[i][j])
      print li
